### ISSUE TITLE
Ports configured as macsuck_no_deviceports get a single MAC added against them on each run of macsuck

### ISSUE THREAD

**Issue by rkerr** [netdisco/netdisco#1234]
For some time I've been noticing MAC entries accumulating on ports that are configured not to be macsucked with macsuck_no_deviceports.

Running in debug I find the config is being picked up ok and netdisco realises the ports are to be excluded:

`[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck Trk1 - port suppressed by macsuck_no_deviceports
`

And it appears to be suppressing the entries:

```
[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck xx:xx:xx:xx:xx:ed - port Trk1 is suppressed by config - skipping.
[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck xx:xx:xx:xx:xx:94 - port Trk1 is suppressed by config - skipping.
[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck xx:xx:xx:xx:xx:40 - port Trk1 is suppressed by config - skipping.
[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck xx:xx:xx:xx:xx:46 - port Trk1 is suppressed by config - skipping.
[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck xx:xx:xx:xx:xx:6e - port Trk1 is suppressed by config - skipping.
[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck xx:xx:xx:xx:xx:4a - port Trk1 is suppressed by config - skipping.
[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck xx:xx:xx:xx:xx:6b - port Trk1 is suppressed by config - skipping.
[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck xx:xx:xx:xx:xx:6a - port Trk1 is suppressed by config - skipping.
[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck xx:xx:xx:xx:xx:ea - port Trk1 is suppressed by config - skipping.
```

However it then goes on to store a single node on the port:

`[17762] 2024-07-12 14:54:09 debug  [x.x.x.x] macsuck - port Trk1 vlan 1 : 1 nodes
`

Turning on SQL tracing I see: 

```
UPDATE node
  SET active = true, time_last = to_timestamp( 1720796048.683122 ) ::timestamp
WHERE mac = 'xx:xx:xx:xx:xx:ea' AND port = 'Trk1' AND switch = 'x.x.x.x' AND vlan = '1'
```

The MAC being stored appears to be the last one on the list of suppressed MAC addresses, and this is consistent across multiple runs and multiple devices. It's always the last 'suppressed' device that gets stored in the database against the port.

**Comment by rkerr** [comment]
Having a look at the sanity_macs function Workers/Plugin/Netdisco/Macsuck/Nodes.pm. There's a section:

```
              if (exists $ignoreport->{$port}) {
                  # stash in the neighborport so that node search works for neighbor
                  # (besides this, neighborport is not used for ignoreport ports)
                  $neighborport->{$port} = [ $vlan, $mac ] if exists $port_macs->{$mac};

                  debug sprintf
                    ' [%s] macsuck %s - port %s is suppressed by config - skipping.',
                    $device->ip, $mac, $port;
                  delete $cache->{$vlan}->{$port}->{$mac};
                  next MAC;
              }
```

Commenting out the $neighborport line seems to stop it storing a single MAC on each ignored port.

I'm not sure I entirely follow the logic here, I think it is intended to store the MAC of a direct neighbour if one is present? however the code as is can't differentiate between a direct neighbour and a device many hops away. The MACs that are being added do appear to be port MACs of other devices in the network, but they are generally not devices that are directly connected to the device being macsucked, so should not be appearing on that port.

**Comment by ollyg** [comment]
thanks @rkerr and thanks for the investigation! should be fixed in b02a713a
